rust/libnewsboat/src/keycombination.rs

```diff
@@ -0,0 +1,343 @@
+        self.control == ControlState::Control
+    }
+
+    pub fn has_alt(&self) -> bool {
+        self.alt == AltState::Alt
+    }
+}
+
+fn alphabetic(input: &str) -> IResult<&str, char> {
+    verify(anychar, |c: &char| is_alphabetic(*c as u8))(input)
+}
+
+fn regular_bindkey(input: &str) -> IResult<&str, KeyCombination> {
+    Ok(("", KeyCombination::new(input.to_owned())))
+}
+
+fn control_bindkey(input: &str) -> IResult<&str, KeyCombination> {
+    let (input, _) = tag("^")(input)?;
+    let (input, key) = alphabetic(input)?;
+    eof(input)?;
+
+    let key_combination = KeyCombination::new(key.to_lowercase().to_string()).with_control();
+    Ok((input, key_combination))
+}
+
+fn shift_bindkey(input: &str) -> IResult<&str, KeyCombination> {
+    let (input, key) = verify(alphabetic, |c: &char| c.is_uppercase())(input)?;
+    eof(input)?;
+
+    let key_combination = KeyCombination::new(key.to_lowercase().to_string()).with_shift();
+    Ok((input, key_combination))
+}
+
+pub fn bindkey(input: &str) -> KeyCombination {
+    let result = alt((shift_bindkey, control_bindkey, regular_bindkey))(input);
+    // Should be save to unwrap because `regular_bindkey` accepts any input
+    let (_, key_combination) = result.unwrap();
+    key_combination
+}
+
+fn control_key_bind(input: &str) -> IResult<&str, KeyCombination> {
+    let (input, _) = tag("^")(input)?;
+    let (input, key) = alphabetic(input)?;
+
+    let key_combination = KeyCombination::new(key.to_lowercase().to_string()).with_control();
+    Ok((input, key_combination))
+}
+
+fn shift_key_bind(input: &str) -> IResult<&str, KeyCombination> {
+    let (input, key) = verify(alphabetic, |c: &char| c.is_uppercase())(input)?;
+
+    let key_combination = KeyCombination::new(key.to_lowercase().to_string()).with_shift();
+    Ok((input, key_combination))
+}
+
+fn combination_key_bind(input: &str) -> IResult<&str, KeyCombination> {
+    let (input, _) = tag("<")(input)?;
+    let (input, modifiers) = many0(alt((tag("C-"), tag("S-"), tag("M-"))))(input)?;
+    let (input, key) = recognize(many0(none_of(">")))(input)?;
+    let (input, _) = tag(">")(input)?;
+
+    let mut key_combination = KeyCombination::new(key.to_owned());
+
+    for modifier in modifiers {
+        match modifier {
+            "C-" => key_combination = key_combination.with_control(),
+            "S-" => key_combination = key_combination.with_shift(),
+            "M-" => key_combination = key_combination.with_alt(),
+            _ => (),
+        }
+    }
+
+    Ok((input, key_combination))
+}
+
+fn single_key_bind(input: &str) -> IResult<&str, KeyCombination> {
+    let (input, key) = anychar(input)?;
+    let key_combination = KeyCombination::new(key.to_string());
+    Ok((input, key_combination))
+}
+
+pub fn bind(input: &str) -> Vec<KeyCombination> {
+    let result = many0(alt((
+        control_key_bind,
+        shift_key_bind,
+        combination_key_bind,
+        single_key_bind,
+    )))(input);
```

Comment:
It might be good to add `let (input, _) = eof(input)?;` to make sure we parsed everything
I will add that after this is reviewed or just before merging (to avoid even more force-pushes and rebuilds)

Reply:
Yes, definitely do that! If we accept a broken config once, we'd have to keep compatibility with it for a long time, so it's better to be restrictive, at least initially.

Reply:
Done